core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -1,158 +1,408 @@
+                    IQueue.empty.enqueue(addToQueue.toList),
+                    IQueue.empty
+                  )
+                else
+                  strategy match {
+                    case BackPressure =>
+                      IO.now(true) -> Surplus(
+                        IQueue.empty[A].enqueue(addToQueue.toList),
+                        IQueue.empty.enqueue(surplusValues -> p)
+                      )
+                    case Sliding =>
+                      val notTaken = addToQueue ++ surplusValues
+                      val toQueue  = capacity.fold(notTaken)(notTaken.takeRight)
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(toQueue.toList),
+                        IQueue.empty
+                      )
+                    case Dropping =>
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(addToQueue.toList),
+                        IQueue.empty
+                      )
+                  }
+
+              completeTakers *> complete -> surplus
+            } else
+              completeTakers *> p.complete(true) -> Deficit(deficitValues)
+        }
+
+      case (p, Surplus(values, putters)) =>
+        val (addToQueue, surplusValues) =
+          capacity.fold((as, Iterable.empty[A]))(c => as.splitAt(c - values.size))
+        if (surplusValues.isEmpty)
+          (p.complete(true), Surplus(values.enqueue(addToQueue.toList), putters))
+        else {
+          strategy match {
+            case BackPressure =>
+              (
+                IO.now(true),
+                Surplus(values.enqueue(addToQueue.toList), putters.enqueue(surplusValues -> p))
+              )
+            case Sliding =>
+              (
+                p.complete(true),
+                capacity.fold(
+                  Surplus(values.enqueue(as.toList), putters)
+                ) { c =>
+                  Surplus(
+                    values
+                      .takeRight(c - as.size)
+                      .enqueue(as.takeRight(c).toList),
+                    putters
+                  )
+                }
+              )
+            case Dropping =>
+              (
+                p.complete(false),
+                Surplus(values.enqueue(addToQueue.toList), putters)
+              )
+          }
+        }
+
+      case (p, state @ Shutdown(errors)) => (p.interrupt0(errors), state)
+    }
+
+    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
+      case (_, p) => p.poll.void <> removePutter(p)
+    }
+
+    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
   }
+
+  /**
+   * Adds a shutdown hook that will be executed when `shutdown` is called.
+   * If the queue is already shutdown, the hook will be executed immediately.
+   */
+  final def onShutdown(io: IO[Nothing, Unit]): IO[Nothing, Unit] =
+    for {
+      state <- ref.get
```

Comment:
This is not safe. Because the state can change between when `ref.get` is called, and when the subsequent logic, including `shutdownHook.modify`, is performed.

To avoid this problem, you need to store the shutdown hook inside the `ref` (the main state of the queue). Basically, all states except `Shutdown` will need to store the shutdown hook and keep passing it around until the end of the queue. `Shutdown` state doesn't need to store the hook because it will be run when the queue is shutdown and then additional calls to `onShutdown` will be run right away.

In general, having more than one ref is not a good idea because you can't change them both in an atomic way.